[FSDP][Replicate] added replicate function that uses FSDP instead of DDP

anshul-si opened the pull request:
**Summary**
Users would like to use Replicate with TP. Currently, the replicate function uses DDP, which has not been maintained resulting in a lack of integration options. Since users can use FSDP with TP, we will make the replicate function use FSDP so that users can use replicate with FSDP. To that end I have created a replicate function that uses FSDP instead of DDP. One blocker that I ran into is that the replicate function has a contract which assigns a module "replicate" attribute in registry. This would mean that fully_shards is_composable requirement would not be satisfied making it impossible to apply fully_shard to a replicate module. The solution to this was to copy the fully_shard function and state and modify it for replicate. In the future, it should be explored making the replicate_state inherit from FSDP_state to get rid of code duplicity. I have attached below the profile tracing of a replicated Net Module.

https://interncache-all.fbcdn.net/manifold/perfetto-artifacts/tree/ui/index.html#!/?url=https://interncache-all.fbcdn.net/manifold/perfetto_internal_traces/tree/shared_trace/anshulsi_270fcc36-194a-42f5-9841-cace984c2132_devgpu263.prn2.facebook.com_1792146.1753232748025155780.pt.trace.json

**Test Case**
1.  pytest test/distributed/_composable/test_replicate_with_fsdp.py

Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #158207




cc @H-Huang @awgu @wanchaol @fegin @fduwjj @wz337 @wconstab @d4l3k @pragupta

sinhaanshul committed:
[FSDP][Replicate] added replicate function that uses FSDP instead of DDP

[ghstack-poisoned]

mori360 commented on torch/distributed/_composable/replicate_with_fsdp.py:
```diff
@@ -0,0 +1,679 @@
+            )
+
+            module.register_forward_pre_hook(_reconstruct_dtensor)
+            module.register_forward_hook(_localize_dtensor)
```
Can we add a unit test to test device mesh is not None in kwargs.

mori360 commented on torch/distributed/_composable/replicate_with_fsdp.py:
```diff
@@ -0,0 +1,679 @@
+    registry = _get_registry(module)
+    if registry is None:
+        return False
+    return "fully_shard" in registry
```
can we combine `is_composable_with_replicate` and `_is_fully_sharded`

mori360 commented on test/distributed/_composable/test_replicate_with_fsdp.py:
```diff
@@ -0,0 +1,209 @@
+            else:
+                self.assertTrue("fully_shard" in _get_registry(layer))
+                for parameter in layer.parameters():
+                    self.assertEqual(parameter.placements, (Shard(dim=0),))
```
Xan we check the root's mesh here?

sinhaanshul committed:
Update on "[FSDP][Replicate] added replicate function that uses FSDP instead of DDP"

**Summary**
Users would like to use Replicate with TP. Currently, the replicate function uses DDP, which has not been maintained resulting in a lack of integration options. Since users can use FSDP with TP, we will make the replicate function use FSDP so that users can use replicate with FSDP. To that end I have created a replicate function that uses FSDP instead of DDP. One blocker that I ran into is that the replicate function has a contract which assigns a module "replicate" attribute in registry. This would mean that fully_shards is_composable requirement would not be satisfied making it impossible to apply fully_shard to a replicate module. The solution to this was to copy the fully_shard function and state and modify it for replicate. In the future, it should be explored making the replicate_state inherit from FSDP_state to get rid of code duplicity. 

**Test Case**
1.  pytest test/distributed/_composable/test_replicate_with_fsdp.py





[ghstack-poisoned]

mori360 commented on test/distributed/_composable/test_replicate_with_fsdp.py:
```diff
@@ -88,6 +90,40 @@ def _init_pg(self):
             store=dist.FileStore(self.file_name, self.world_size),
         )
 
+    def check_sharded_parity(
```
can we import it from fsdp_common?

mori360 commented on test/distributed/_composable/test_replicate_with_fsdp.py:
```diff
@@ -120,15 +156,19 @@ def step_model(model, input, target):
             torch.manual_seed(iteration)
             input = input[torch.randperm(global_batch_size)]
 
+    def run_subtests(self, *args, **kwargs):
```
what's the difference between these 2 `run_subtests`

sinhaanshul committed:
Update on "[FSDP][Replicate] added replicate function that uses FSDP instead of DDP"

**Summary**
Users would like to use Replicate with TP. Currently, the replicate function uses DDP, which has not been maintained resulting in a lack of integration options. Since users can use FSDP with TP, we will make the replicate function use FSDP so that users can use replicate with FSDP. To that end I have created a replicate function that uses FSDP instead of DDP. One blocker that I ran into is that the replicate function has a contract which assigns a module "replicate" attribute in registry. This would mean that fully_shards is_composable requirement would not be satisfied making it impossible to apply fully_shard to a replicate module. The solution to this was to copy the fully_shard function and state and modify it for replicate. In the future, it should be explored making the replicate_state inherit from FSDP_state to get rid of code duplicity. 

**Test Case**
1.  pytest test/distributed/_composable/test_replicate_with_fsdp.py





[ghstack-poisoned]

mori360 commented on test/distributed/_composable/test_replicate_with_fsdp.py:
```diff
@@ -0,0 +1,270 @@
+        return run_subtests(self, *args, **kwargs)
+
+    @skip_if_lt_x_gpu(2)
+    def test_replicate_net_module(self):
```
What's the difference between `test_replicate_transformer` and  `test_replicate_net_module`

sinhaanshul committed:
Update on "[FSDP][Replicate] added replicate function that uses FSDP instead of DDP"

**Summary**
Users would like to use Replicate with TP. Currently, the replicate function uses DDP, which has not been maintained resulting in a lack of integration options. Since users can use FSDP with TP, we will make the replicate function use FSDP so that users can use replicate with FSDP. To that end I have created a replicate function that uses FSDP instead of DDP. One blocker that I ran into is that the replicate function has a contract which assigns a module "replicate" attribute in registry. This would mean that fully_shards is_composable requirement would not be satisfied making it impossible to apply fully_shard to a replicate module. The solution to this was to copy the fully_shard function and state and modify it for replicate. In the future, it should be explored making the replicate_state inherit from FSDP_state to get rid of code duplicity. 

**Test Case**
1.  pytest test/distributed/_composable/test_replicate_with_fsdp.py





[ghstack-poisoned]

sinhaanshul committed:
Update on "[FSDP][Replicate] added replicate function that uses FSDP instead of DDP"

**Summary**
Users would like to use Replicate with TP. Currently, the replicate function uses DDP, which has not been maintained resulting in a lack of integration options. Since users can use FSDP with TP, we will make the replicate function use FSDP so that users can use replicate with FSDP. To that end I have created a replicate function that uses FSDP instead of DDP. One blocker that I ran into is that the replicate function has a contract which assigns a module "replicate" attribute in registry. This would mean that fully_shards is_composable requirement would not be satisfied making it impossible to apply fully_shard to a replicate module. The solution to this was to copy the fully_shard function and state and modify it for replicate. In the future, it should be explored making the replicate_state inherit from FSDP_state to get rid of code duplicity. 

**Test Case**
1.  pytest test/distributed/_composable/test_replicate_with_fsdp.py





[ghstack-poisoned]

mori360 commented on test/distributed/_composable/test_replicate_with_fsdp.py:
```diff
@@ -0,0 +1,233 @@
+            elif i % 3 == 1:
+                self.assertTrue("fully_shard" in _get_registry(layer))
+                for parameter in layer.parameters():
+                    self.assertEqual(parameter.placements, (Shard(dim=0),))
```
can we the check for the layers that i%3==2?

mori360 commented on test/distributed/_composable/test_replicate_with_fsdp.py:
```diff
@@ -0,0 +1,233 @@
+        for layer in layers:
+            fully_shard(layer)
+
+        replicate_model = fully_shard(replicate_model)
```
compare the model training with replicate

mori360 replied:
can we have another unit test to train with replicate and tp
https://github.com/pytorch/pytorch/blob/da05b7fb94fa6382c43e165a525a76d8ae62cadd/test/distributed/_composable/test_composability/test_2d_composability.py#L117-L137

mori360 commented on torch/distributed/_composable/replicate_with_fsdp.py:
```diff
@@ -0,0 +1,655 @@
+
+@contract(state_cls=_ReplicateState)
+def replicate(
+    module: nn.Module,
```
can we add the ignored_module back, we can utilize ignored_params in fully_shard to do the ignore

mori360 commented on torch/distributed/_composable/replicate_with_fsdp.py:
```diff
@@ -0,0 +1,655 @@
+        self._modules_to_run_forward: set[nn.Module] = set()
+        # ``False`` when user set reshard_after_forward
+        # through ``fully_shard`` or ``set_reshard_after_forward``
+        self._auto_reshard_after_forward: Optional[bool] = True
```
we can remove `self._auto_reshard_after_forward` here

mori360 commented on torch/distributed/_composable/replicate_with_fsdp.py:
```diff
@@ -0,0 +1,655 @@
+        )
+
+    mesh = mesh or _init_default_fully_shard_mesh()
+    if mesh.ndim != 2:
```
if user input 1d mesh(replicate mesh), we should init the 2d (replicate, shard) mesh and run replicate

sinhaanshul committed:
Update on "[FSDP][Replicate] added replicate function that uses FSDP instead of DDP"

**Summary**
Users would like to use Replicate with TP. Currently, the replicate function uses DDP, which has not been maintained resulting in a lack of integration options. Since users can use FSDP with TP, we will make the replicate function use FSDP so that users can use replicate with FSDP. To that end I have created a replicate function that uses FSDP instead of DDP. One blocker that I ran into is that the replicate function has a contract which assigns a module "replicate" attribute in registry. This would mean that fully_shards is_composable requirement would not be satisfied making it impossible to apply fully_shard to a replicate module. The solution to this was to copy the fully_shard function and state and modify it for replicate. In the future, it should be explored making the replicate_state inherit from FSDP_state to get rid of code duplicity. 

**Test Case**
1.  pytest test/distributed/_composable/test_replicate_with_fsdp.py





[ghstack-poisoned]

weifengpy commented on torch/distributed/_composable/replicate_with_fsdp.py:
```diff
@@ -0,0 +1,399 @@
+
+_ROOT_MODULE_PREFIX = ""
+
+logger = logging.getLogger("torch.distributed.fsdp.fully_shard")
```
logger = logging.getLogger("torch.distributed._composable.replicate_with_fsdp")

sinhaanshul committed:
Update on "[FSDP][Replicate] added replicate function that uses FSDP instead of DDP"

**Summary**
Users would like to use Replicate with TP. Currently, the replicate function uses DDP, which has not been maintained resulting in a lack of integration options. Since users can use FSDP with TP, we will make the replicate function use FSDP so that users can use replicate with FSDP. To that end I have created a replicate function that uses FSDP instead of DDP. One blocker that I ran into is that the replicate function has a contract which assigns a module "replicate" attribute in registry. This would mean that fully_shards is_composable requirement would not be satisfied making it impossible to apply fully_shard to a replicate module. The solution to this was to copy the fully_shard function and state and modify it for replicate. In the future, it should be explored making the replicate_state inherit from FSDP_state to get rid of code duplicity. I have attached below the profile tracing of a replicated Net Module.

https://interncache-all.fbcdn.net/manifold/perfetto-artifacts/tree/ui/index.html#!/?url=https://interncache-all.fbcdn.net/manifold/perfetto_internal_traces/tree/shared_trace/anshulsi_4fe55b66-5e56-4667-ab2d-aeb773d9e799_devgpu263.prn2.facebook.com_1304563.1753231986593603819.pt.trace.json

**Test Case**
1.  pytest test/distributed/_composable/test_replicate_with_fsdp.py





[ghstack-poisoned]

sinhaanshul committed:
Update on "[FSDP][Replicate] added replicate function that uses FSDP instead of DDP"

**Summary**
Users would like to use Replicate with TP. Currently, the replicate function uses DDP, which has not been maintained resulting in a lack of integration options. Since users can use FSDP with TP, we will make the replicate function use FSDP so that users can use replicate with FSDP. To that end I have created a replicate function that uses FSDP instead of DDP. One blocker that I ran into is that the replicate function has a contract which assigns a module "replicate" attribute in registry. This would mean that fully_shards is_composable requirement would not be satisfied making it impossible to apply fully_shard to a replicate module. The solution to this was to copy the fully_shard function and state and modify it for replicate. In the future, it should be explored making the replicate_state inherit from FSDP_state to get rid of code duplicity. I have attached below the profile tracing of a replicated Net Module.

https://interncache-all.fbcdn.net/manifold/perfetto-artifacts/tree/ui/index.html#!/?url=https://interncache-all.fbcdn.net/manifold/perfetto_internal_traces/tree/shared_trace/anshulsi_270fcc36-194a-42f5-9841-cace984c2132_devgpu263.prn2.facebook.com_1792146.1753232748025155780.pt.trace.json

**Test Case**
1.  pytest test/distributed/_composable/test_replicate_with_fsdp.py





[ghstack-poisoned]

sinhaanshul committed:
Update on "[FSDP][Replicate] added replicate function that uses FSDP instead of DDP"

**Summary**
Users would like to use Replicate with TP. Currently, the replicate function uses DDP, which has not been maintained resulting in a lack of integration options. Since users can use FSDP with TP, we will make the replicate function use FSDP so that users can use replicate with FSDP. To that end I have created a replicate function that uses FSDP instead of DDP. One blocker that I ran into is that the replicate function has a contract which assigns a module "replicate" attribute in registry. This would mean that fully_shards is_composable requirement would not be satisfied making it impossible to apply fully_shard to a replicate module. The solution to this was to copy the fully_shard function and state and modify it for replicate. In the future, it should be explored making the replicate_state inherit from FSDP_state to get rid of code duplicity. I have attached below the profile tracing of a replicated Net Module.

https://interncache-all.fbcdn.net/manifold/perfetto-artifacts/tree/ui/index.html#!/?url=https://interncache-all.fbcdn.net/manifold/perfetto_internal_traces/tree/shared_trace/anshulsi_270fcc36-194a-42f5-9841-cace984c2132_devgpu263.prn2.facebook.com_1792146.1753232748025155780.pt.trace.json

**Test Case**
1.  pytest test/distributed/_composable/test_replicate_with_fsdp.py





[ghstack-poisoned]

sinhaanshul committed:
Update on "[FSDP][Replicate] added replicate function that uses FSDP instead of DDP"

**Summary**
Users would like to use Replicate with TP. Currently, the replicate function uses DDP, which has not been maintained resulting in a lack of integration options. Since users can use FSDP with TP, we will make the replicate function use FSDP so that users can use replicate with FSDP. To that end I have created a replicate function that uses FSDP instead of DDP. One blocker that I ran into is that the replicate function has a contract which assigns a module "replicate" attribute in registry. This would mean that fully_shards is_composable requirement would not be satisfied making it impossible to apply fully_shard to a replicate module. The solution to this was to copy the fully_shard function and state and modify it for replicate. In the future, it should be explored making the replicate_state inherit from FSDP_state to get rid of code duplicity. I have attached below the profile tracing of a replicated Net Module.

https://interncache-all.fbcdn.net/manifold/perfetto-artifacts/tree/ui/index.html#!/?url=https://interncache-all.fbcdn.net/manifold/perfetto_internal_traces/tree/shared_trace/anshulsi_270fcc36-194a-42f5-9841-cace984c2132_devgpu263.prn2.facebook.com_1792146.1753232748025155780.pt.trace.json

**Test Case**
1.  pytest test/distributed/_composable/test_replicate_with_fsdp.py





[ghstack-poisoned]
